AAUTH-559 spi for conditionless responses

ThePetrov opened the pull request:
Create an SPI for conditionless response handlers. Those handlers will be fired after every successful processing of a SAML response which does not have conditions in it. Those handlers will receive a list of resolved issuers to make identification of the offending IdPs easier.

When debugging the code I noticed that the issuers are not determined correctly - the size of the collection is incorrectly checked. I fixed that problem and added a matching test case.

ThePetrov committed:
NONE: correctly check for response issuer

ThePetrov committed:
AAUTH-559: add pluggable SPI for conditionless responses

dserkowski commented on core/src/main/java/com/onelogin/saml2/authn/SamlResponse.java:
```diff
@@ -343,6 +348,12 @@ public boolean isValid(String requestId) {
 	}
 
+	private void handleConditionlessResponses(List<String> issuers) throws XPathExpressionException {
+		if (getConditions().getLength() == 0) {
```
Will it work well for requests that do not send the condition tag at all? I mean won't `getConditions()` return null?
Just for double-checking

ThePetrov replied:
It won't, this is exercised by `com.onelogin.saml2.test.authn.AuthnResponseTest#testIsInValidConditionsWithoutConditionsValidation`

dserkowski commented on core/src/main/java/com/onelogin/saml2/authn/SamlResponse.java:
```diff
@@ -715,7 +730,7 @@ public Boolean checkOneAuthnStatement() throws XPathExpressionException {
 		String value;
 		NodeList responseIssuer = Util.query(samlResponseDocument, "/samlp:Response/saml:Issuer");
-		if (responseIssuer.getLength() > 1) {
+		if (responseIssuer.getLength() > 0) {
```
nice catch!

dserkowski commented on core/src/test/java/com/onelogin/saml2/test/authn/AuthnResponseTest.java:
```diff
@@ -1811,7 +1863,7 @@ public void testIsInValidIssuer() throws IOException, Error, XPathExpressionExce
 		samlResponse = new SamlResponse(settings, newHttpRequest(samlResponseEncoded));
 		assertFalse(samlResponse.isValid());
-		assertEquals("No Signature found. SAML Response rejected", samlResponse.getError());
+		assertEquals("Invalid issuer in the Assertion/Response. Was 'http://invalid.isser.example.com/', but expected 'http://idp.example.com/'", samlResponse.getError());
```
typo? 
```suggestion
		assertEquals("Invalid issuer in the Assertion/Response. Was 'http://invalid.issuer.example.com/', but expected 'http://idp.example.com/'", samlResponse.getError());
```

ThePetrov replied:
Yes, but in the data file - you can inspect `data/responses/invalids/invalid_issuer_message.xml.base64` (you'll need to base64 decode it). If I correct the typo then the test will fail ;)

dserkowski commented on core/src/main/java/com/onelogin/saml2/authn/SamlResponse.java:
```diff
@@ -343,6 +348,12 @@ public boolean isValid(String requestId) {
 
+	private void handleConditionlessResponses(List<String> issuers) throws XPathExpressionException {
+		if (getConditions().getLength() == 0) {
+			settings.getConditionlessResponseHandlers().forEach(h -> h.handleConditionlessResponse(issuers));
```
I do not know the whole context behind this. Handlers you are introducing are something that can be injected by apps consuming this lib, aren't it? I assume that after merging this, you will create additional PR in the AAUTH plugin to inject a custom handler. Did I understand it correctly?

ThePetrov replied:
That's correct. Defining such callbacks in AAUTH gives us a lot of more possibilities when it comes to what we can call in such a handler, here we're limited to java-saml's classpath.

hysterus1 replied:
Yeah, this is not targeted to solve the problem, but provide a way so we can fix it on our end :)

ThePetrov committed:
AAUTH-559: adjust SPI to handle amount of authnstatements different than 1

- rename wantConditionsPresent to compatibility mode, invert
- don't check if there's exactly one authnstatement if compatibility mode enabled
- replace multiple handlers with single handler
- add test case for multiple authn statements
- correct issuers/audience in no_authnstatement.xml.base64

ThePetrov commented on core/src/main/java/com/onelogin/saml2/authn/SamlResponse.java:
```diff
@@ -259,8 +262,10 @@ public boolean isValid(String requestId) {
 				// Validate AuthnStatement element exists and is unique
-				if (!this.checkOneAuthnStatement()) {
-					throw new ValidationError("The Assertion must include an AuthnStatement element", ValidationError.WRONG_NUMBER_OF_AUTHSTATEMENTS);
+				if (!settings.isCompatibilityMode()) {
```
This is also a change from the current behavior. This should be present according to the spec, but let's be as compatible as possible for the time being
>The set of one or more bearer assertions MUST contain at least one <AuthnStatement> thatreflects the authentication of the principal to the identity provider. Multiple <AuthnStatement>elements MAY be included, but the semantics of multiple statements is not defined by this profile
- Profiles Section 4.1.4.2

ThePetrov commented on core/src/test/resources/data/responses/invalids/no_authnstatement.xml.base64:
```diff
@@ -1 +1 @@
-PHNhbWxwOlJlc3BvbnNlIHhtbG5zOnNhbWxwPSJ1cm46b2FzaXM6bmFtZXM6dGM6U0FNTDoyLjA6cHJvdG9jb2wiIHhtbG5zOnNhbWw9InVybjpvYXNpczpuYW1lczp0YzpTQU1MOjIuMDphc3NlcnRpb24iIElEPSJwZngwNWYzY2UxMC0xNjE1LWYzZWEtYTk4OC02MGUzODBiMzI5OWYiIFZlcnNpb249IjIuMCIgSXNzdWVJbnN0YW50PSIyMDE0LTAyLTE5VDAxOjM3OjAxWiIgRGVzdGluYXRpb249Imh0dHBzOi8vZXhhbXBsZS5jb20vbmV3b25lbG9naW4vZGVtbzEvaW5kZXgucGhwP2FjcyIgSW5SZXNwb25zZVRvPSJPTkVMT0d ...
\ No newline at end of file
+PHNhbWxwOlJlc3BvbnNlIHhtbG5zOnNhbWxwPSJ1cm46b2FzaXM6bmFtZXM6dGM6U0FNTDoyLjA6cHJvdG9jb2wiIHhtbG5zOnNhbWw9InVybjpvYXNpczpuYW1lczp0YzpTQU1MOjIuMDphc3NlcnRpb24iIElEPSJwZngwNWYzY2UxMC0xNjE1LWYzZWEtYTk4OC02MGUzODBiMzI5OWYiIFZlcnNpb249IjIuMCIgSXNzdWVJbnN0YW50PSIyMDE0LTAyLTE5VDAxOjM3OjAxWiIgRGVzdGluYXRpb249Imh0dHA6Ly9sb2NhbGhvc3Q6ODA4MC9qYXZhLXNhbWwtanNwc2FtcGxlL2Fjcy5qc3AiIEluUmVzcG9uc2VUbz0iT05FTE9HSU5 ...
```
This data file had an invalid issue and audience restriction, I wanted to have a test which processes such a response and treats it as valid if compatibility mode is on, so I fixed it

ThePetrov committed:
AAUTH-559: renamed handleConditionlessResponse

jpodeszwik commented on core/src/main/java/com/onelogin/saml2/authn/SamlResponse.java:
```diff
@@ -1,15 +1,15 @@
 package com.onelogin.saml2.authn;
 
-import java.io.IOException;
```
Nit: would it be better to keep original sorting? This would help to avoid conflicts in case we decide to merge some upstream changes.

ThePetrov replied:
The original project doesn't define import order, mine's the IntelliJ default. I can try guessing the current import order, but there's a risk that I'll get it wrong. Every file that we modify means a risk that the import order will be modified again. I'd leave it as is - in the worst case one can ignore all importing changes in the merge view and then run organize imports (IIRC it can be executed in the merge view)

jpodeszwik commented on core/src/main/java/com/onelogin/saml2/authn/SamlResponse.java:
```diff
@@ -343,6 +351,17 @@ public boolean isValid(String requestId) {
+		final int amountOfAuthnStatements = getAuthnStatements().getLength();
+		if (amountOfConditions == 0 || amountOfAuthnStatements != 1) {
+			final CompatibilityModeViolationHandler handler = settings.getCompatibilityModeViolationHandler();
+			if (handler != null) {
```
Is it common situation that this will be null? If no, then maybe consider logging a warning here.

ThePetrov replied:
In our case it will be always set. I do like the idea of verifying though - I think a test on the plugin side which checks if we configure the handler correctly would be beneficial. Thanks!
